### PR TITLE
refactor: move the logic of override host selecting out from load balancing policy

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: refactor: move the logic of override host selecting out from load balancing policy
Additional Description:

The override host selecting is implemented in every load balancing policy for now. The code is fragmented and it's not friendly for the hierarchical LBs like subset lb.

This refactor moved the logic of override host selecting out from load balancing policy to make sure this feature is load balancing policy independent.
And this may make the code cleaner and easier to maintain.

Risk Level: mid.
Testing: unit.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @markdroth This PR will make the `override_host_statues` policy independent.

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -1721,10 +1722,33 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::httpConnPoolIsIdle(
   }
 }
 
+HostConstSharedPtr ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::chooseHost(
+    LoadBalancerContext* context) {
+  auto corss_priority_host_map = priority_set_.crossPriorityHostMap();
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc:1727]:
```suggestion
  auto cross_priority_host_map = priority_set_.crossPriorityHostMap();
```

Same elsewhere.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Looks like a nice cleanup, thanks.

/wait

**Comment by markdroth** [comment]
Great to see this change!  Thank you so much for doing this!

**Comment by markdroth** [comment]
This is part of #20634.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**